src/sp/it/pl/util/system/Environment.kt

```diff
@@ -47,96 +47,105 @@ fun copyToSysClipboard(s: String?) = copyToSysClipboard(DataFormat.PLAIN_TEXT, s
+                    .directory(parentDirOrRoot)
+                    .start()
+                    .apply(then)
 
-        ok<Exception>()
-    } catch (e: IOException) {
-        logger.warn(e) { "Failed to launch program" }
-        Try.error<Void, Exception>(e)
+            Try.ok<Process, Exception>(process)
+        } catch (e: IOException) {
+            logger.error(e) { "Failed to launch program" }
+            Try.error<Process, Exception>(e)
+        }
     }
 }
 
 /**
- * Runs a command in new background thread as a process and executes an action (on it) right after it launches.
- * This allows process monitoring or waiting for it to end.
+ * Runs a command in new background thread as a separate process on a new thread and executes an action (on it) right
+ * after it launches.
  *
  * @param command see [Runtime.exec]
+ * @param then block taking the program's process as parameter executing if the program executes
+ * @return success if the program is executed or error if it is not, irrespective of if and how the program finishes
  */
 @JvmOverloads
-fun runCommand(command: String, then: ((Process) -> Unit)? = null) {
-    runNew {
+fun runCommand(command: String, then: (Process) -> Unit = {}): Fut<Try<Process, Exception>> {
+    return runNew {
         try {
-            val p = Runtime.getRuntime().exec(command)
-            then?.invoke(p)
+            val process = Runtime.getRuntime().exec(command).apply(then)
+            Try.ok<Process, Exception>(process)
         } catch (e: IOException) {
             logger.error(e) { "Error running command '$command'" }
+            Try.error<Process, Exception>(e)
         }
     }
 }
 
-/**
- * Browse the file in the systems file browser by opening the parent directory and selecting it.
- * If it is unsupported, it will try to [open] the parent directory instead.
- */
-fun File.browse() {
-    if (Os.WINDOWS.isCurrent) {
-        openWindowsExplorerAndSelect()
-    } else {
-        // Would be nice if this was widely supported, but it isn't
-        if (Desktop.Action.BROWSE_FILE_DIR.isSupportedOrWarn()) {
-            Desktop.getDesktop().browseFileDirectory(this)
-        } else {
-            parentDirOrRoot.open()
-        }
-    }
-}
+/** Equivalent to [URI.browse] for the URI denoting this file. */
+fun File.browse() = toURI().browse()
 
 /**
- * Browse uri:
- * - directory/file -> calls [File.browse]
- * - url -> opens it in system browser
+ * Browse this uri, in order:
+ * - if it is directory/file -> browse the file in the system file browser by opening the parent directory and selecting the file.
+ * If this is unsupported, [open] is called on the parent directory.
+ * - if it is url -> open the url in system browser
  *
  * On some platforms the operation may be unsupported. In that case this method is a no-op.
  */
 fun URI.browse() {
     logger.info { "Browsing uri=$this" }
     runNew {
-        val browse = toFileOrNull()?.browse()
```

Comment:
could you please not undo half of my changes without saying a word about it?

Reply:
I think I wrote somewhere about how your solution was blocking call that can not stay public and making it non blocking would end up spawning multiple threads in some situations and about File.browse() being specialization of URI.browse(). Well, after I rewrote it...

I understand the reaction and this is indeed your PR, but it was the only possible solution to the problem and I already had it open in the editor. I did that because asking you to do it would have the same result, but with communication overhead. I'm fine with that, but I do not wish to burden you more than necessary.

Frankly, I do not like the used ifNull { }.ifNotNull { } so it is not like I'm happy with the current code.

Reply:
If you wish we can agree on sctrict owner-only PR modification rule. It will lead to more comments and longer process I think, but I'm fine with it.

Reply:
but the way you wrote it converts a File to a URI and then back when you browse. That doesn't sound like a sane thing to me, since the way I wrote it seems perfectly valid. If the blocking is a problem, then add a `runNew` call and maybe you can somehow check if you are already in a runNew call to avoid blocking. Or just make it all blocking ^^

Reply:
1 Ah I see File -> URI -> File is the problem. It is not elegant, but is correct. From the point of view of File.browse we only do File -> URI (as we should). The fact that URI checks whether it denotes a file by ```File(this)```, an actual conversion, is an implementation detail which may change (for instance we could check if uri has 'file:' scheme, but I am no expert so I felt this way is the most reliable).
2 This will not become a performance problem
3 Even in absence of the threading issue, File.browse should still delegate to URI.browse for multiple reasons. You appear to disagree. Your solution would mean that URI.browse() would depend on File.browse() which would mean the two methods would be intrinsically related - and they would be have to be made sure to stay synchronized. Having only one method that does provides an actual browse implementation fixes this problem. This is not hypothetical or far-fetched scenario. It is important to lower code maintenance and most of all absolutely prevent nay possible 'mutation'. Now File.browse is always guaranteed to work the same way as URI.browse, at all times in 100% of cases. This is reflected by javadoc of File.browse merely linking to URI.browse. And developer also can easily understand what File.browse does.

4 the proposed solution of runNew with a Thread check is fragile solution leading to bugs and maintenance nightmares, because it intorduces hidden dependency of two implementations: 1 thread check in File.browse and execution context in URI.browse. In order to be consistent, we would also need to add this check to all memthods in ENvironment or cause inconsistent behavior between these methods. Lastly, the proposed check can never work reliably: you mean to denote the thread with a special name or something? What does it mean that we are in a 'new' thread? Which thread does not mind to be blocked and which doesnt? Who decides? This is entirely wrong way to go about this.

5 The code was written before coroutines were brought into the project. I'm still not sure we want to use coroutines themselves so I havent used them anywhere, I think the Fut is much easier on the mind and fills 90% of the needs. That being said launch {} could indeed solve this issue, but that would require making the function suspending and that would screw up the API for java as well as cause more problems in places where coroutines are not exactly compatible/expected/needed/wanted.

I do think Environment is one of the prime subjects for considering using coroutines. That would however be a refactoring that is largely unnecessary and as such the work should go elsewhere.